runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceView.java

```diff
@@ -232,9 +225,35 @@ public void write(IToken token, Object data) throws OverwriteException {
      */
     public @Nonnull ILogData read(long address) {
         if (!runtime.getParameters().isCacheDisabled()) {
-            ILogData data;
             try {
-                data = readCache.get(address);
+                // The VersionLockedObject and the Transaction layer will generate
+                // undoRecord(s) during a transaction commit, or object sync. These
+                // undo records are stored in transient fields and are not persisted.
+                // A missing undo record can cause a NoRollbackException, thus forcing
+                // a complete object rebuild that generates a "scanning" behavior
+                // which affects the LRU window. In essence, affecting other cache users
+                // and making the VersionLockedObject very sensitive to caching behavior.
+                // A concrete example of this would be unsynchronized readers/writes:
+                // 1. Thread A starts replicating write1
+                // 2. Thread B discovers the write (via stream tail query) and
+                //    tries to read write1
+                // 3. Thread B's read results in a cache miss and the reader thread
+                //    starts loading the value into the cache
+                // 4. Thread A completes its write and caches it with undo records
+                // 5. Thread B finishes loading and caches the loaded value replacing
+                //    the cached value from step 4 (i.e. loss of undo records computed
+                //    by thread A)
+                ILogData data = readCache.getIfPresent(address);
+                if (data == null) {
+                    // Loading a value without the cache loader can result in
+                    // redundant loading calls (i.e. multiple threads try to
+                    // load the same value), but currently a redundant RPC
+                    // is much cheaper than the cost of a NoRollBackException, therefore
+                    // this trade-off is reasonable
+                    final ILogData loadedVal = fetch(address);
+                    data = readCache.get(address, () -> loadedVal);
+                    return data;
```

Comment:
We are missing to return the data if it is present in the cache, right?

Reply:
If the mapping exists in the cache guava will return the old value and loadedVal will be ignored.
Essentially, line 254 will be 
`data = existing mapping (not loadedVal)`

Reply:
I was referring to line 246, if it is already present you can return it... otherwise It'll exercise the fetch and go to the log unit again.

Reply:
Good catch!